src/common/eigf.h

```diff
@@ -0,0 +1,301 @@
+
+  float ming = 10000000.0f;
+  float maxg = 0.0f;
+  float minm = 10000000.0f;
+  float maxm = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+dt_omp_firstprivate(guide, mask, Ndim) \
+  schedule(simd:static) aligned(guide, mask:64) \
+  reduction(max:maxg, maxm)\
+  reduction(min:ming, minm)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float pixelg = guide[k];
+    const float pixelm = mask[k];
+    if(pixelg < ming) ming = pixelg;
+    if(pixelg > maxg) maxg = pixelg;
+    if(pixelm < minm) minm = pixelm;
+    if(pixelm > maxm) maxm = pixelm;
+  }
+
+  dt_gaussian_t *g = dt_gaussian_init(width, height, 1, &maxg, &ming, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, guide, blurred_guide);
+  dt_gaussian_free(g);
+
+  g = dt_gaussian_init(width, height, 1, &maxm, &minm, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, mask, blurred_mask);
+  dt_gaussian_free(g);
+
+  float mingg = 10000000.0f;
+  float maxgg = 0.0f;
+  float mingm = 10000000.0f;
+  float maxgm = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(guide, mask, blurred_guide, blurred_mask, guide_x_guide, guide_x_mask, Ndim) \
+  schedule(simd:static) aligned(guide, mask, blurred_guide, blurred_mask, guide_x_guide, guide_x_mask:64) \
+  reduction(max:maxgg, maxgm)\
+  reduction(min:mingg, mingm)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float deviation = guide[k] - blurred_guide[k];
+    const float squared_deviation = deviation * deviation;
+    guide_x_guide[k] = squared_deviation;
+    if(squared_deviation < mingg) mingg = squared_deviation;
+    if(squared_deviation > maxgg) maxgg = squared_deviation;
+    const float cov = (guide[k] - blurred_guide[k]) * (mask[k] - blurred_mask[k]);
+    guide_x_mask[k] = cov;
+    if(cov < mingm) mingm = cov;
+    if(cov > maxgm) maxgm = cov;
+  }
+
+  g = dt_gaussian_init(width, height, 1, &maxgg, &mingg, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, guide_x_guide, guide_variance);
+  dt_gaussian_free(g);
+
+  g = dt_gaussian_init(width, height, 1, &maxgm, &mingm, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, guide_x_mask, guide_mask_covariance);
+  dt_gaussian_free(g);
+
+  float mina = 10000000.0f;
+  float minb = 10000000.0f;
+  float minw = 10000000.0f;
+  float maxa = 0.0f;
+  float maxb = 0.0f;
+  float maxw = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(guide, mask, blurred_guide, guide_variance, blurred_mask, guide_mask_covariance, a, b, weights, Ndim, feathering) \
+  schedule(simd:static) aligned(guide, mask, blurred_guide, guide_variance, blurred_mask, guide_mask_covariance, a, b, weights:64) \
+  reduction(max:maxa, maxb, maxw)\
+  reduction(min:mina, minb, minw)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    // allow max 2EV below average for normalization (avoid noise issues)
+    const float pixelg = fmaxf(guide[k], fmaxf(blurred_guide[k] * 0.25f, 1E-6));
```

Comment:
Here, we can see the change that give better noise tolerance:
we normalize using the maximum between the pixel value and the local average - 2EV.
In other word, we consider any pixel which is below local average - 2EV to be exactly 2EV below average.
As both the pixel value and 0.25 * average are changed the same way when changing exposure, the filter remains exposure-invariant